misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,73 @@
+"""AMIA paper 2024: Statistics
+
+Resources
+- GH issue: https://github.com/jhu-bids/TermHub/issues/631
+- Paper: https://livejohnshopkins-my.sharepoint.com/:w:/g/personal/sgold15_jh_edu/EXbuxwBsb0pPgFfgPYjR6qgBlJ1gmpqsMPYJ3vn6Y_HnYw?e=4%3ApEDbMe&fromShare=true&at=9&CID=d4663c40-1ebe-db59-8cf5-2bc091d45692"""
+import os
+import sys
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List
+
+import pandas as pd
+
+THIS_DIR = Path(os.path.dirname(__file__))
+PROJECT_ROOT = THIS_DIR.parent.parent
+sys.path.insert(0, str(PROJECT_ROOT))
+from backend.routes.db import usage_query
+
+
+OUTDIR = THIS_DIR / 'output'
+CSV_PATH = OUTDIR / 'usage.csv'
+
+
+def setup():
+    """Setup any environmental prerequisites"""
+    if not OUTDIR.exists():
+        OUTDIR.mkdir()
+
+
+def get_data() -> pd.DataFrame:
+    """Get data to be analyzed"""
+    if not os.path.exists(CSV_PATH):
+        # JSON_FILE = '/Users/joeflack4/Desktop/vs-hub-usage-report.json'
+        # with open(JSON_FILE, 'r') as file:
+        #     d = json.load(file)
+        # Fetch
+        data: List[Dict] = usage_query()
+        df = pd.DataFrame(data)
+        # Save
+        t0 = datetime.now()
+        df.to_csv(CSV_PATH, index=False)
+        print(f'Wrote {len(df)} records to CSV in n seconds: {(datetime.now() - t0).seconds}')
+    else:
+        # Read
+        t0 = datetime.now()
+        df = pd.read_csv(CSV_PATH)
+        print(f'Read {len(df)} records from CSV in n seconds: {(datetime.now() - t0).seconds}')
+    return df
+
+
+def preprocess(df: pd.DataFrame, verbose=True) -> pd.DataFrame:
```

Comment:
## Dev IP preprocessing
Maybe I should've added this to `usage_query()`, but I added it to my `amia_paper_2024/stats.py`.

I'm surprised that it's showing this few records for you, and 0 records for me:
> ```
> Filtered out n records based on IP addresses of each dev:
>   - Siggie: 62
>   - Joe: 0
> ```

Reply:
Also, surprisingly, all of these entries are cases where `api_call_group_id = NULL`. If I remove all of those cases, it doesn't show any records for dev IP addresses.

Reply:
FYI I added more ideas etc to a [comment in the paper](https://livejohnshopkins-my.sharepoint.com/:w:/r/personal/sgold15_jh_edu/Documents/JBI-paper-2022-revision/[codesethub-amia-2023-12-draft.docx](https://livejohnshopkins-my.sharepoint.com/:w:/r/personal/sgold15_jh_edu/Documents/JBI-paper-2022-revision/codesethub-amia-2023-12-draft.docx?d=w00c7ee766f6c4f4a8057e03d88d1eaa8&csf=1&web=1&e=oWRpIB&nav=eyJjIjoxNzg1MjgyNTV9)?d=w00c7ee766f6c4f4a8057e03d88d1eaa8&csf=1&web=1&e=oWRpIB&nav=eyJjIjoxNzg1MjgyNTV9).